File: app/src/main/java/com/github/premnirmal/ticker/network/ApiModule.java
package com.github.premnirmal.ticker.network;

import android.content.Context;
import android.content.SharedPreferences;

import com.github.premnirmal.ticker.RefreshReceiver;
import com.github.premnirmal.ticker.RxBus;
import com.github.premnirmal.ticker.UpdateReceiver;
import com.github.premnirmal.ticker.model.HistoryProvider;
import com.github.premnirmal.ticker.model.IHistoryProvider;
import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.ticker.model.StocksProvider;
import com.github.premnirmal.ticker.portfolio.AddPositionActivity;
import com.github.premnirmal.ticker.portfolio.EditPositionActivity;
import com.github.premnirmal.ticker.portfolio.GraphActivity;
import com.github.premnirmal.ticker.portfolio.PortfolioFragment;
import com.github.premnirmal.ticker.portfolio.RearrangeActivity;
import com.github.premnirmal.ticker.portfolio.TickerSelectorActivity;
import com.github.premnirmal.ticker.settings.SettingsActivity;
import com.github.premnirmal.ticker.widget.RemoteStockViewAdapter;
import com.github.premnirmal.ticker.widget.StockWidget;
import com.github.premnirmal.tickerwidget.R;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit.RestAdapter;

/**
 * Created by premnirmal on 12/21/14.
 */

@Module(
        injects = {
                TickerSelectorActivity.class,
                RemoteStockViewAdapter.class,
                SettingsActivity.class,
                StockWidget.class,
                UpdateReceiver.class,
                RefreshReceiver.class,
                GraphActivity.class,
                PortfolioFragment.class,
                RearrangeActivity.class,
                AddPositionActivity.class,
                EditPositionActivity.class
        },
        complete = false,
        library = true
)
public class ApiModule {

    private StocksApi stocksApi;
    private SuggestionApi suggestionApi;

    @Provides
    @Singleton
    StocksApi provideStocksApi(YahooFinance yahooFinance) {
        if (stocksApi == null) {
            stocksApi = new StocksApi(yahooFinance);
        }
        return stocksApi;
    }

    @Provides
    @Singleton
    YahooFinance provideYahooFinance(Context context) {
        final RestAdapter restAdapter = new RestAdapter.Builder()
                .setEndpoint(context.getString(R.string.yahoo_endpoint))
                .build();
        final YahooFinance yahooFinance = restAdapter.create(YahooFinance.class);
        return yahooFinance;
    }

//    @Provides
//    @Singleton
//    GoogleFinance provideGoogleFinance(Context context) {
//        final RestAdapter restAdapter = new RestAdapter.Builder()
//                .setEndpoint(context.getString(R.string.google_endpoint))
//                .setConverter(new GStockConverter())
//                .build();
//        final GoogleFinance googleFinance = restAdapter.create(GoogleFinance.class);
//        return googleFinance;
//    }

    @Provides
    @Singleton
    SuggestionApi provideSuggestionsApi(Context context) {
        if (suggestionApi == null) {
            final RestAdapter restAdapter = new RestAdapter.Builder()
                    .setEndpoint(context.getString(R.string.suggestions_endpoint))
                    .setConverter(new StupidYahooWrapConverter())
                    .build();
            suggestionApi = restAdapter.create(SuggestionApi.class);
        }
        return suggestionApi;
    }

    @Provides
    @Singleton
    IStocksProvider provideStocksProvider(Context context, StocksApi stocksApi, RxBus bus, SharedPreferences sharedPreferences) {
        return new StocksProvider(stocksApi, bus, context, sharedPreferences);
    }

    @Provides
    @Singleton
    IHistoryProvider provideHistoryProvider(Context context, StocksApi stocksApi) {
        return new HistoryProvider(stocksApi, context);
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/network/ApiModule.kt
package com.github.premnirmal.ticker.network

import android.content.Context
import android.content.SharedPreferences
import com.github.premnirmal.ticker.RxBus
import com.github.premnirmal.ticker.model.HistoryProvider
import com.github.premnirmal.ticker.model.IHistoryProvider
import com.github.premnirmal.ticker.model.IStocksProvider
import com.github.premnirmal.ticker.model.StocksProvider
import com.github.premnirmal.tickerwidget.BuildConfig
import com.github.premnirmal.tickerwidget.R
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

/**
 * Created on 3/3/16.
 */
@Module
class ApiModule {

  lateinit var stocksApi: StocksApi
  lateinit var suggestionApi: SuggestionApi


  companion object {
    internal val CONNECTION_TIMEOUT: Long = 20000
    internal val READ_TIMEOUT: Long = 20000
  }

  @Provides @Singleton
  internal fun provideHttpClient(context: Context, bus: RxBus): OkHttpClient {
    val logger = HttpLoggingInterceptor()
    logger.level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE)
    val okHttpClient = OkHttpClient.Builder()
        .addInterceptor(NoCacheRequestInterceptor(context, bus))
        .readTimeout(READ_TIMEOUT, TimeUnit.MILLISECONDS)
        .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
        .build()
    return okHttpClient
  }

  @Provides @Singleton
  internal fun provideStocksApi(yahooFinance: YahooFinance,
      googleFinance: GoogleFinance): StocksApi {
    stocksApi = StocksApi(yahooFinance, googleFinance)
    return stocksApi
  }

  @Provides @Singleton
  internal fun provideYahooFinance(context: Context, okHttpClient: OkHttpClient): YahooFinance {
    val Retrofit = Retrofit.Builder()
        .client(okHttpClient)
        .baseUrl(context.getString(R.string.yahoo_endpoint))
        .(
            if (BuildConfig.DEBUG) Retrofit.LogLevel.FULL else Retrofit.LogLevel.NONE)
        .build()
    val yahooFinance = Retrofit.create(YahooFinance::class.java)
    return yahooFinance
  }

  @Provides @Singleton
  internal fun provideGoogleFinance(context: Context, okHttpClient: OkHttpClient): GoogleFinance {
    val Retrofit: Retrofit = Retrofit.Builder()
        .client(okHttpClient)
        .baseUrl(context.getString(R.string.google_endpoint))
        .setConverter(GStockConverter())
        .build()
    val googleFinance: GoogleFinance = Retrofit.create(GoogleFinance::class.java)
    return googleFinance
  }

  @Provides @Singleton
  internal fun provideSuggestionsApi(context: Context, okHttpClient: OkHttpClient): SuggestionApi {
    val Retrofit = Retrofit.Builder()
        .client(okHttpClient)
        .baseUrl(context.getString(R.string.suggestions_endpoint))
        .setConverter(StupidYahooWrapConverter())
        .build()
    suggestionApi = Retrofit.create(SuggestionApi::class.java)
    return suggestionApi
  }

  @Provides @Singleton
  internal fun provideStocksProvider(context: Context, stocksApi: StocksApi, bus: RxBus,
      sharedPreferences: SharedPreferences): IStocksProvider {
    return StocksProvider(stocksApi, bus, context, sharedPreferences)
  }

  @Provides @Singleton
  internal fun provideHistoryProvider(context: Context, stocksApi: StocksApi): IHistoryProvider {
    return HistoryProvider(stocksApi, context)
  }

}
